### PR TITLE
fix https://github.com/huggingface/accelerate/issues/2627

### LINKED ISSUE #2627: accelerator prepared dataloader doesn't shuffle
### System Info

```Shell
- `Accelerate` version: 0.25.0
- Platform: Linux-5.4.143.bsk.7-amd64-x86_64-with-glibc2.31
- Python version: 3.9.2
- Numpy version: 1.24.3
- PyTorch version (GPU?): 2.1.0+cu118 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 2015.23 GB
- GPU type: NVIDIA A100-SXM4-80GB
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Minimal Reproducible Example
```
import torch
from accelerate import Accelerator
from datasets import Dataset

dataset = Dataset.from_list([{"idx": i} for i in range(1000)])

accelerator = Accelerator()

train_dataloader = torch.utils.data.DataLoader(
    dataset,
    shuffle=True,
    sampler=None,
    batch_size=4,
    num_workers=0,
)

train_dataloader = accelerator.prepare(train_dataloader)

if accelerator.is_main_process:
    print("===================================================================")
    print(f"train_dataloader: {train_dataloader}")
    print(f"train_dataloader.sampler: {train_dataloader.sampler}")
    print(f"train_dataloader.batch_sampler: {train_dataloader.batch_sampler}")
    print(f"train_dataloader.batch_sampler.batch_sampler: {train_dataloader.batch_sampler.batch_sampler}")
    print(f"train_dataloader.batch_sampler.batch_sampler.sampler: {train_dataloader.batch_sampler.batch_sampler.sampler}")
    print("===================================================================")

for epoch in range(3):
    train_dataloader.set_epoch(epoch)
    for iteration, data in enumerate(train_dataloader):
        if accelerator.is_main_process:
            print(f"Rank: [{accelerator.process_index}/{accelerator.num_processes}] - Epoch: {epoch} - Iteration: {iteration} - Data: {data}")
        break
```

### Command

```
accelerate launch main.py
```

### Results

```
===================================================================
train_dataloader: <accelerate.data_loader.DataLoaderShard object at 0x7f19ca08a1f0>
train_dataloader.sampler: <torch.utils.data.sampler.SequentialSampler object at 0x7f19ca08a2b0>
train_dataloader.batch_sampler: <accelerate.data_loader.BatchSamplerShard object at 0x7f19ca08ad30>
train_dataloader.batch_sampler.batch_sampler: <torch.utils.data.sampler.BatchSampler object at 0x7f19ca083280>
train_dataloader.batch_sampler.batch_sampler.sampler: <accelerate.data_loader.SeedableRandomSampler object at 0x7f19ca08a8b0>
===================================================================
Rank: [0/8] - Epoch: 0 - Iteration: 0 - Data: {'idx': tensor([855, 780, 107, 766], device='cuda:0')}
Rank: [0/8] - Epoch: 1 - Iteration: 0 - Data: {'idx': tensor([855, 780, 107, 766], device='cuda:0')}
Rank: [0/8] - Epoch: 2 - Iteration: 0 - Data: {'idx': tensor([855, 780, 107, 766], device='cuda:0')}
````

:bug::bug::bug: **No Shuffle at different epochs**

### Expected behavior

The first iteration at different epoch should sample different data. 

After fix this bug in https://github.com/huggingface/accelerate/pull/2628, the result becomes:
```
Rank: [0/8] - Epoch: 0 - Iteration: 0 - Data: {'idx': tensor([552, 326, 904, 718], device='cuda:0')}
Rank: [0/8] - Epoch: 1 - Iteration: 0 - Data: {'idx': tensor([702, 969,  19, 494], device='cuda:0')}
Rank: [0/8] - Epoch: 2 - Iteration: 0 - Data: {'idx': tensor([517, 403, 850, 746], device='cuda:0')}
```

### PR DESCRIPTION
**Pull request by ShoufaChen**
# What does this PR do?

Fixes #2627

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@muellerzr

### CODE REVIEW

**Review by muellerzr — COMMENTED** [review]
Thanks for the PR! Any chance you could add a test in `src/accelerate/test_utils/scripts/test_script.py` for this?

**Comment by ShoufaChen** [comment]
Hi, @muellerzr ,

Thanks for reviewing. Is the Minimal Reproducible Example I posted in https://github.com/huggingface/accelerate/issues/2627 enough?

**Comment by muellerzr** [comment]
Yes!

**Comment by ShoufaChen** [comment]
Then do I need move this Minimal Reproducible Example to the test script?

**Comment by muellerzr** [comment]
Yes, as a function and make sure it gets called. You can see other examples of how this is done inside that file :)

**Comment by ShoufaChen** [comment]
Hi @muellerzr ,

After a careful check, I believe https://github.com/huggingface/accelerate/commit/ef68b4655cbb2cc8f322946c877c0e0ae1bf6806 has fixed this bug by moving
```

    if isinstance(sampler, SeedableRandomSampler) and use_seedable_sampler:
        if sampler_is_batch_sampler:
            dataloader.sampler.sampler = sampler
        else:
            dataloader.batch_sampler.sampler = sampler
```

from before the position of `DataLoaderShard` initialization to after that.

In this way, the `train_dataloader.batch_sampler` would have `sampler` attribute.

Feel free to close this PR.

**Comment by ShoufaChen** [comment]
In summary, this bug was unintentionally introduced in the great feature on https://github.com/huggingface/accelerate/commit/7843286f2e1c50735d259fbc0084a7f1c85e00e3 (Oct 24, 2023) and fixed in https://github.com/huggingface/accelerate/commit/ef68b4655cbb2cc8f322946c877c0e0ae1bf6806 (Feb 13, 2024). 

Therefore, accelerate versions from [v0.24.0](https://github.com/huggingface/accelerate/releases/tag/v0.24.0) to [v0.27.1](https://github.com/huggingface/accelerate/releases/tag/v0.27.1) would have this problem.

From [v0.27.2](https://github.com/huggingface/accelerate/releases/tag/v0.27.2) it would work fine as shown in the commit history [here](https://github.com/huggingface/accelerate/commits/5eb7ab9f6afb0fed38cb54741e243180d6f8bc5d/).

Just putting this information here in case it might be helpful. 🤖

**Comment by ShoufaChen** [comment]
Hi @muellerzr ,

I believe that a DataLoader lacking a shuffle property is a serious bug and may not be easily identifiable by users. Is there any possibility of issuing a warning or applying a patch to these affected versions?

I'm aiming to help prevent difficulties and save time for all users.

Thank you.

**Comment by muellerzr** [comment]
> Is there any possibility of issuing a warning or applying a patch to these affected versions?

I'll see what we can do, otherwise the open issue/searching should be more than enough (most people live on the bleeding edge of accelerate we've found)

**Comment by ShoufaChen** [comment]
Thank you for considering this issue. I agree that people can search for and find this issue if they are aware that it relates to the shuffling feature of the DataLoader. However, it may not be immediately apparent that shuffling is the problem if they experience undesirable performance like accuracy drop.
